File: January2022/Java/AllElementsInTwoBinarySearchTrees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> getAllElements(TreeNode root1, TreeNode root2) {
        // Do inorder of both trees and store them in 2 lists. Then merge 2 sorted lists.
        // Inorder traversal of BST will give sorted list.
        List<Integer> tree1 = new ArrayList<>();
        List<Integer> tree2 = new ArrayList<>();
        inorder(root1, tree1);
        inorder(root2, tree2);
        
        List<Integer> res = new ArrayList<>();
        merge(tree1, tree2, res);
        return res;
    }
    
    private void inorder(TreeNode root, List<Integer> tree) {
        if (root == null)
            return;
        inorder(root.left, tree);
        tree.add(root.val);
        inorder(root.right, tree);
    }
    
    private void merge(List<Integer> tree1, List<Integer> tree2, List<Integer> res) {
        int l1 = tree1.size();
        int l2 = tree2.size();
        
        int i = 0, j = 0;
        while (i < l1 && j < l2) {
            if (tree1.get(i) <= tree2.get(j)) {
                res.add(tree1.get(i));
                i += 1;
            }
            else {
                res.add(tree2.get(j));
                j += 1;
            }
            
        }
        while (i < l1) {
            res.add(tree1.get(i));
            i += 1;
        }
        while (j < l2) {
            res.add(tree2.get(j));
            j += 1;
        }
        
    }
}

File: January2022/Swift/All Elements in Two Binary Search Trees.swift
// Author: Deepal
// Problem: All Elements in Two Binary Search Trees
// Link: https://leetcode.com/problems/all-elements-in-two-binary-search-trees/
// Category: Medium
// DataStructure: Binary Search Tree

func getAllElements(_ root1: TreeNode?, _ root2: TreeNode?) -> [Int] {
    var result1 = [Int]()
    var result2 = [Int]()
    inOrderTraverse(root: root1, result: &result1)
    inOrderTraverse(root: root2, result: &result2)
    let result = mergeTwoArray(array1: &result1, array2: &result2)
    return result
}

private func inOrderTraverse(root: TreeNode?, result: inout [Int]) {
    if root != nil {
        inOrderTraverse(root: root?.left, result: &result)
        result.append(root?.val ?? 0)
        inOrderTraverse(root: root?.right, result: &result)
    }
}

private func mergeTwoArray(array1: inout [Int], array2: inout [Int]) -> [Int] {
    var result = [Int](repeating: 0, count: array1.count + array2.count)
    
    var index1 = 0
    var index2 = 0
    var index3 = 0
    
    while index1 < array1.count && index2 < array2.count {
        if array1[index1] < array2[index2] {
            result[index3] = array1[index1]
            index3 += 1
            index1 += 1
        } else {
            result[index3] = array2[index2]
            index3 += 1
            index2 += 1
        }
    }
    
    while index1 < array1.count {
        result[index3] = array1[index1]
        index3 += 1
        index1 += 1
    }
    
    while index2 < array2.count {
        result[index3] = array2[index2]
        index3 += 1
        index2 += 1
    }
    return result
}

// TC: O(M+N) 
// SC: O(M+N)

let allElementsInTwoBSTOutput = Solution().getAllElements(TreeNode(2, TreeNode(1), TreeNode(4)),
                                                          TreeNode(1, TreeNode(0), TreeNode(3)))
print(allElementsInTwoBSTOutput)
